File: backend/app/services/cluster_service.py
"""
ApeAI — Cluster Service

Manages feedback clusters — groups of related feedback items
that represent a common theme or user request.

Clusters are central to the pipeline: they flow through states
from 'new' → 'clustered' → 'brd_generated' → ... → 'tickets_created'.
"""

import logging
from typing import Dict, Any, List, Optional

from backend.app.db.supabase_client import get_supabase_client

logger = logging.getLogger(__name__)


async def create_cluster(
    title: str,
    summary: Optional[str] = None,
    feedback_ids: Optional[List[str]] = None,
    confidence_score: float = 0.0,
) -> Dict[str, Any]:
    """
    Create a new feedback cluster.
    
    Args:
        title: Descriptive title for the cluster theme.
        summary: Brief summary of the cluster.
        feedback_ids: List of feedback UUIDs to include.
        confidence_score: AI confidence score (0-100).
        
    Returns:
        The created cluster row.
    """
    db = get_supabase_client()

    # Create the cluster
    cluster_data = {
        "title": title,
        "summary": summary,
        "confidence_score": confidence_score,
        "feedback_count": len(feedback_ids) if feedback_ids else 0,
        "status": "new",
    }

    result = db.table("clusters").insert(cluster_data).execute()

    if not result.data:
        raise RuntimeError("Failed to create cluster")

    cluster = result.data[0]
    cluster_id = cluster["id"]

    logger.info(f"✅ Cluster created: {cluster_id} — '{title}'")

    # Link feedback items to the cluster
    if feedback_ids:
        links = [
            {"cluster_id": cluster_id, "feedback_id": fid}
            for fid in feedback_ids
        ]
        db.table("cluster_feedback").insert(links).execute()
        logger.info(f"   Linked {len(feedback_ids)} feedback items")

    return cluster


async def get_cluster(cluster_id: str, include_feedback: bool = True) -> Dict[str, Any]:
    """
    Get a cluster by ID, optionally including its feedback items.
    
    Args:
        cluster_id: UUID of the cluster.
        include_feedback: Whether to include feedback items.
        
    Returns:
        Cluster data with optional feedback items.
    """
    db = get_supabase_client()

    result = db.table("clusters") \
        .select("*") \
        .eq("id", cluster_id) \
        .single() \
        .execute()

    if not result.data:
        raise ValueError(f"Cluster not found: {cluster_id}")

    cluster = result.data

    if include_feedback:
        # Get linked feedback items
        links = db.table("cluster_feedback") \
            .select("feedback_id, similarity_score") \
            .eq("cluster_id", cluster_id) \
            .execute()

        feedback_items = []
        if links.data:
            fids = [l["feedback_id"] for l in links.data]
            scores = {l["feedback_id"]: l.get("similarity_score") for l in links.data}

            feedback = db.table("feedback") \
                .select("*") \
                .in_("id", fids) \
                .execute()

            for f in (feedback.data or []):
                f["similarity_score"] = scores.get(f["id"])
                feedback_items.append(f)

        cluster["feedback_items"] = feedback_items

    return cluster


async def list_clusters(
    status: Optional[str] = None,
    limit: int = 50,
    offset: int = 0,
) -> Dict[str, Any]:
    """
    List clusters with optional status filter.
    
    Args:
        status: Filter by pipeline status (e.g., 'new', 'approved').
        limit: Number of clusters to return.
        offset: Pagination offset.
        
    Returns:
        Dict with clusters list and total count.
    """
    db = get_supabase_client()

    query = db.table("clusters") \
        .select("*", count="exact") \
        .order("created_at", desc=True) \
        .range(offset, offset + limit - 1)

    if status:
        query = query.eq("status", status)

    result = query.execute()

    total = result.count if result.count is not None else len(result.data or [])

    return {
        "clusters": result.data or [],
        "total": total,
    }


async def update_cluster(
    cluster_id: str,
    title: Optional[str] = None,
    summary: Optional[str] = None,
    status: Optional[str] = None,
    confidence_score: Optional[float] = None,
) -> Dict[str, Any]:
    """
    Update a cluster's fields.
    
    Only non-None fields are updated.
    """
    db = get_supabase_client()

    update_data = {}
    if title is not None:
        update_data["title"] = title
    if summary is not None:
        update_data["summary"] = summary
    if status is not None:
        update_data["status"] = status
    if confidence_score is not None:
        update_data["confidence_score"] = confidence_score

    if not update_data:
        raise ValueError("No fields to update")

    result = db.table("clusters") \
        .update(update_data) \
        .eq("id", cluster_id) \
        .execute()

    if not result.data:
        raise ValueError(f"Cluster not found: {cluster_id}")

    logger.info(f"✅ Cluster updated: {cluster_id}")
    return result.data[0]


async def delete_cluster(cluster_id: str) -> bool:
    """
    Delete a cluster and all its links.
    
    Cascade delete handles cluster_feedback and documents.
    """
    db = get_supabase_client()

    result = db.table("clusters") \
        .delete() \
        .eq("id", cluster_id) \
        .execute()

    if result.data:
        logger.info(f"🗑️  Cluster deleted: {cluster_id}")
        return True

    raise ValueError(f"Cluster not found: {cluster_id}")


async def add_feedback_to_cluster(
    cluster_id: str,
    feedback_ids: List[str],
    similarity_scores: Optional[List[float]] = None,
) -> Dict[str, Any]:
    """
    Add feedback items to an existing cluster.
    
    Args:
        cluster_id: UUID of the cluster.
        feedback_ids: List of feedback UUIDs to add.
        similarity_scores: Optional similarity scores for each item.
        
    Returns:
        Summary with count of items added.
    """
    db = get_supabase_client()

    links = []
    for i, fid in enumerate(feedback_ids):
        link = {"cluster_id": cluster_id, "feedback_id": fid}
        if similarity_scores and i < len(similarity_scores):
            link["similarity_score"] = similarity_scores[i]
        links.append(link)

    result = db.table("cluster_feedback").insert(links).execute()

    # Update feedback count on the cluster
    count_result = db.table("cluster_feedback") \
        .select("id", count="exact") \
        .eq("cluster_id", cluster_id) \
        .execute()

    new_count = count_result.count if count_result.count is not None else len(count_result.data or [])
    db.table("clusters").update({"feedback_count": new_count}).eq("id", cluster_id).execute()

    logger.info(f"✅ Added {len(feedback_ids)} items to cluster {cluster_id}")
    return {"added": len(result.data or []), "total_in_cluster": new_count}


async def remove_feedback_from_cluster(
    cluster_id: str,
    feedback_ids: List[str],
) -> Dict[str, Any]:
    """Remove feedback items from a cluster."""
    db = get_supabase_client()

    for fid in feedback_ids:
        db.table("cluster_feedback") \
            .delete() \
            .eq("cluster_id", cluster_id) \
            .eq("feedback_id", fid) \
            .execute()

    # Update feedback count
    count_result = db.table("cluster_feedback") \
        .select("id", count="exact") \
        .eq("cluster_id", cluster_id) \
        .execute()

    new_count = count_result.count if count_result.count is not None else len(count_result.data or [])
    db.table("clusters").update({"feedback_count": new_count}).eq("id", cluster_id).execute()

    logger.info(f"🗑️  Removed {len(feedback_ids)} items from cluster {cluster_id}")
    return {"removed": len(feedback_ids), "total_in_cluster": new_count}


async def get_cluster_stats() -> Dict[str, Any]:
    """Get cluster statistics grouped by status."""
    db = get_supabase_client()

    try:
        result = db.table("clusters").select("status").execute()

        by_status: Dict[str, int] = {}
        for row in (result.data or []):
            s = row.get("status", "unknown")
            by_status[s] = by_status.get(s, 0) + 1

        return {
            "total": len(result.data or []),
            "by_status": by_status,
        }
    except Exception as e:
        logger.error(f"Failed to get cluster stats: {e}")
        return {"total": 0, "by_status": {}, "error": str(e)}

File: backend-node/src/services/clusterService.ts
/**
 * ApeAI — Cluster Service
 *
 * Manages feedback clusters — groups of related feedback items.
 * Mirrors Python services/cluster_service.py exactly.
 */

import { getSupabaseClient } from '../db/supabase';

type DbRow = Record<string, unknown>;

export async function createCluster(params: {
  title: string;
  summary?: string;
  feedbackIds?: string[];
  confidenceScore?: number;
  userId?: string;
}): Promise<DbRow> {
  const db = getSupabaseClient();
  const { title, summary, feedbackIds, confidenceScore = 0.0, userId } = params;

  const clusterData: DbRow = {
    title,
    summary: summary ?? null,
    confidence_score: confidenceScore,
    feedback_count: feedbackIds?.length ?? 0,
    status: 'new',
  };
  if (userId) clusterData['user_id'] = userId;

  const { data, error } = await db.from('clusters').insert(clusterData).select().single();
  if (error || !data) throw new Error(`Failed to create cluster: ${error?.message}`);

  const cluster = data as DbRow;
  const clusterId = cluster['id'] as string;
  console.info(`✅ Cluster created: ${clusterId} — '${title}'`);

  if (feedbackIds && feedbackIds.length > 0) {
    const links = feedbackIds.map((fid) => ({ cluster_id: clusterId, feedback_id: fid }));
    await db.from('cluster_feedback').insert(links);
    console.info(`   Linked ${feedbackIds.length} feedback items`);
  }

  return cluster;
}

export async function getCluster(
  clusterId: string,
  includeFeedback = true,
): Promise<DbRow> {
  const db = getSupabaseClient();

  const { data, error } = await db
    .from('clusters')
    .select('*')
    .eq('id', clusterId)
    .single();

  if (error || !data) throw new Error(`Cluster not found: ${clusterId}`);

  const cluster = data as DbRow;

  if (includeFeedback) {
    const { data: links } = await db
      .from('cluster_feedback')
      .select('feedback_id, similarity_score')
      .eq('cluster_id', clusterId);

    const feedbackItems: DbRow[] = [];
    if (links && links.length > 0) {
      const fids = links.map((l: DbRow) => l['feedback_id']);
      const scores: Record<string, unknown> = {};
      for (const l of links as DbRow[]) {
        scores[l['feedback_id'] as string] = l['similarity_score'];
      }

      const { data: feedbackData } = await db
        .from('feedback')
        .select('*')
        .in('id', fids);

      for (const f of (feedbackData ?? []) as DbRow[]) {
        feedbackItems.push({ ...f, similarity_score: scores[f['id'] as string] });
      }
    }

    cluster['feedback_items'] = feedbackItems;
  }

  return cluster;
}

export async function listClusters(params: {
  status?: string;
  userId?: string;
  limit?: number;
  offset?: number;
}): Promise<{ clusters: DbRow[]; total: number }> {
  const db = getSupabaseClient();
  const { status, userId, limit = 50, offset = 0 } = params;

  let query = db
    .from('clusters')
    .select('*', { count: 'exact' })
    .order('created_at', { ascending: false })
    .range(offset, offset + limit - 1);

  if (status) query = query.eq('status', status);
  if (userId) query = query.eq('user_id', userId);

  const { data, error, count } = await query;
  if (error) throw new Error(`Failed to list clusters: ${error.message}`);

  return {
    clusters: (data ?? []) as DbRow[],
    total: count ?? (data ?? []).length,
  };
}

export async function updateCluster(
  clusterId: string,
  updates: {
    title?: string;
    summary?: string;
    status?: string;
    confidenceScore?: number;
  },
): Promise<DbRow> {
  const db = getSupabaseClient();
  const updateData: DbRow = {};

  if (updates.title !== undefined) updateData['title'] = updates.title;
  if (updates.summary !== undefined) updateData['summary'] = updates.summary;
  if (updates.status !== undefined) updateData['status'] = updates.status;
  if (updates.confidenceScore !== undefined) updateData['confidence_score'] = updates.confidenceScore;

  if (Object.keys(updateData).length === 0) throw new Error('No fields to update');

  const { data, error } = await db
    .from('clusters')
    .update(updateData)
    .eq('id', clusterId)
    .select()
    .single();

  if (error || !data) throw new Error(`Cluster not found: ${clusterId}`);
  console.info(`✅ Cluster updated: ${clusterId}`);
  return data as DbRow;
}

export async function deleteCluster(clusterId: string): Promise<void> {
  const db = getSupabaseClient();

  const { data, error } = await db
    .from('clusters')
    .delete()
    .eq('id', clusterId)
    .select();

  if (error) throw new Error(`Failed to delete cluster: ${error.message}`);
  if (!data || (data as DbRow[]).length === 0) throw new Error(`Cluster not found: ${clusterId}`);
  console.info(`🗑️  Cluster deleted: ${clusterId}`);
}

export async function addFeedbackToCluster(
  clusterId: string,
  feedbackIds: string[],
  similarityScores?: number[],
): Promise<{ added: number; total_in_cluster: number }> {
  const db = getSupabaseClient();

  const links = feedbackIds.map((fid, i) => ({
    cluster_id: clusterId,
    feedback_id: fid,
    similarity_score: similarityScores?.[i] ?? null,
  }));

  const { data } = await db.from('cluster_feedback').insert(links).select();

  const { count } = await db
    .from('cluster_feedback')
    .select('id', { count: 'exact' })
    .eq('cluster_id', clusterId);

  const newCount = count ?? 0;
  await db.from('clusters').update({ feedback_count: newCount }).eq('id', clusterId);

  console.info(`✅ Added ${feedbackIds.length} items to cluster ${clusterId}`);
  return { added: (data ?? []).length, total_in_cluster: newCount };
}

export async function removeFeedbackFromCluster(
  clusterId: string,
  feedbackIds: string[],
): Promise<{ removed: number; total_in_cluster: number }> {
  const db = getSupabaseClient();

  for (const fid of feedbackIds) {
    await db
      .from('cluster_feedback')
      .delete()
      .eq('cluster_id', clusterId)
      .eq('feedback_id', fid);
  }

  const { count } = await db
    .from('cluster_feedback')
    .select('id', { count: 'exact' })
    .eq('cluster_id', clusterId);

  const newCount = count ?? 0;
  await db.from('clusters').update({ feedback_count: newCount }).eq('id', clusterId);

  console.info(`🗑️  Removed ${feedbackIds.length} items from cluster ${clusterId}`);
  return { removed: feedbackIds.length, total_in_cluster: newCount };
}

export async function getClusterStats(
  userId?: string,
): Promise<{ total: number; by_status: Record<string, number> }> {
  const db = getSupabaseClient();

  let query = db.from('clusters').select('status');
  if (userId) query = query.eq('user_id', userId);

  const { data, error } = await query;

  if (error) {
    console.error('Failed to get cluster stats:', error);
    return { total: 0, by_status: {} };
  }

  const byStatus: Record<string, number> = {};
  for (const row of (data ?? []) as DbRow[]) {
    const s = (row['status'] as string) ?? 'unknown';
    byStatus[s] = (byStatus[s] ?? 0) + 1;
  }

  return { total: (data ?? []).length, by_status: byStatus };
}
